File: apps/web/src/@types/campaign.ts
import { Contact } from "./contact"
import { Template } from "./template"

export const CAMPAIGN_STATUS = {
    "Draft": "DRAFT",
    "Pending": "PENDING",
    "Processing": "PROCESSING",
    "Sent": "SENT",
    "Failed": "FAILED"
} as const

export const CAMPAIGN_SEND_TYPE = {
    "Immediate": "IMMEDIATE",
    "Scheduled": "SCHEDULED",
} as const

export type CampaignStatus = (typeof CAMPAIGN_STATUS)[keyof typeof CAMPAIGN_STATUS]

export type CampaignSendType = (typeof CAMPAIGN_SEND_TYPE)[keyof typeof CAMPAIGN_SEND_TYPE]

export type Campaign = {
    id: string
    name: string
    description: string | null
    sendType: CampaignSendType
    scheduleAt: string | null
    status: CampaignStatus

    template: Template
    contacts: Contact[]
}

File: rem-server/src/main/java/server/rem/entities/Campaign.java
package server.rem.entities;

import java.time.Instant;
import java.util.Set;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;

import jakarta.persistence.*;
import lombok.*;
import server.rem.enums.CampaignSendType;
import server.rem.enums.CampaignStatus;

@Entity
@Table(name = "campaigns")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Campaign extends Base {
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "business_id", nullable = false)
    private Business business;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "template_id", nullable = false)
    private Template template;

    @Column(name = "name", nullable = false)
    private String name;

    @Column(name = "description", nullable = true)
    private String description;

    @Column(name = "send_type", nullable = false)
    @Enumerated(EnumType.STRING)
    private CampaignSendType sendType;

    @Column(name = "schedule_at", nullable = true)
    private Instant scheduleAt;

    @ManyToMany(fetch = FetchType.EAGER, cascade = CascadeType.ALL)
    @JoinTable(name = "campaign_contact", joinColumns = @JoinColumn(name = "campaign_id", referencedColumnName = "id"), inverseJoinColumns = @JoinColumn(name = "contact_id", referencedColumnName = "id"))
    @JsonIgnoreProperties({ "business", "customerGroup", "tag", "campaigns" })
    private Set<Contact> contacts;

    @Column(name = "status", nullable = false)
    @Enumerated(EnumType.STRING)
    @Builder.Default
    private CampaignStatus status = CampaignStatus.PENDING;
}